### PR TITLE
Integration test

### PR DESCRIPTION
**Pull request by c-nixon**
This PR adds a simple example integration test for the agent CLI.

It creates a temp dir, configures the agent to watch the temp dir, starts the agent, creates some files and checks that the agent successfully runs.

It does not currently validate that the expected events happen, only that it does not crash as long as a valid API key is present.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -13,10 +13,22 @@ pipeline {
     }
     stages {
         stage('Test') {
+            environment {
+                CREDS_FILE = credentials('pipeline-e2e-creds')
+                LOGDNA_HOST = "logs.use.stage.logdna.net"
```
Comment (jakedipity) [inline Jenkinsfile:18]: Not sure we want to consider the staging environment stable enough to use for reliable testing. We can isolate the agent completely and use a mock ingester  that tracks what lines have come through. I believe we already have a yaml for that

File: Makefile [hunk 2]
```diff
@@ -19,7 +19,7 @@ DOCKER_PUBLIC_IMAGE := docker.io/logdna/logdna-agent
 DOCKER_IBM_IMAGE := icr.io/ext/logdna-agent
 
 RUST_COMMAND := $(DOCKER_DISPATCH) $(RUST_IMAGE)
-HADOLINT_COMMAND := $(DOCKER_DISPATCH) $(HADOLINT_IMAGE)
+HADOLINT_COMMAND := $(DOCKER_DISPATCH) $(HADOLINT_IMAGE) ""
```
Comment (jakedipity) [inline Makefile:22]: This change seems a little mysterious

File: bin/tests/cli.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+use tempfile::tempdir;
+
+use std::fs::File;
+use std::io::Write;
+use std::process::Command;
+use std::thread;
+
+#[test]
+fn api_key_missing() {
+    let mut cmd = Command::cargo_bin("logdna-agent").unwrap();
```
Comment (jakedipity) [inline bin/tests/cli.rs:13]: We seem to instantiate this object in many places. Does it make sense to create it once and have it available for the entire program?

File: Makefile (hunk 2 above)
Comment (c-nixon) [reply Makefile:22]: Yeh, it's a bit awkward, I can move it to where we invoke the hadolint command to make it more obvious, it's just the equivalent of passing a null argument, unfortunately bash doesn't have named args.

File: Jenkinsfile (hunk 1 above)
Comment (c-nixon) [reply Jenkinsfile:18]: I was going to build that functionality into a future version of the test binary, I didn't know we had a mock ingester.

I used staging as that's what Test suggested. I'm not _too_ concerned about the investees being available in staging, but I agree having something capture the lines and validate delivery would be better here.

I would still like to test that we successfully write to the real ingesters somewhere, maybeit should like in a separate test stage that doesn't automatically fail the build, or checks the api is available and skips if it isn't?

File: bin/tests/cli.rs (hunk 3 above)
Comment (c-nixon) [reply bin/tests/cli.rs:13]: It's the initial state of a mutable builder object which we configure differently in each test, see the next line vs it's equivalent in the api_key_present test. Each of the methods takes a &mut, so to reuse it we'd need to clone it each time before we called any methods and unfortunately it doesn't impl Clone.

We _could_ factor it out into a function to reduce the boilerplate a little?

Comment (jakedipity) [reply bin/tests/cli.rs:13]: I should have payed closer attention. I'm assuming the builder object is modified in the steps below in which case my suggestion doesn't make sense. Wrapping a function around it won't save much of any boilerplate... so it's probably fine to leave as is

**Review by jakedipity — APPROVED** [review]
I think what's here looks good. I'm a bit hesitant as the testing seems partially complete - we don't validate lines have been ingested. I think it would be okay to tackle this in a later update

**Review by jakedipity — DISMISSED** [review]
Looks good to me, before you merge can you squash the commits into one with a tidied up commit message

File: bin/tests/cli.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    thread::sleep(std::time::Duration::from_secs(1));
+
+    handle.kill().unwrap();
+    let mut output = String::new();
+
+    let stderr_ref = handle.stderr.as_mut().unwrap();
+    stderr_ref.read_to_string(&mut output).unwrap();
+
+    // Check that the agent logs that it has sent lines from each file
+    assert!(predicate::str::contains(&format!(
```
Comment (c-nixon) [inline bin/tests/cli.rs:84]: Tada, it now actually checks things.

**Review by jakedipity — DISMISSED** [review]
LGTM

**Pull request merged**